### PR TITLE
Mongodb shard uris, disallow missing database

### PR DESCRIPTION
**Pull request by pimeys**
We copy whatever uri parsing MongoDB crate does until they make their api public.

Closes: https://github.com/prisma/prisma/issues/11684
Closes: https://github.com/prisma/prisma/issues/11873
Closes: https://github.com/prisma/prisma/issues/11909

### CODE REVIEW

File: libs/mongodb-client/Cargo.toml [hunk 1]
```diff
@@ -6,4 +6,7 @@ edition = "2021"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-mongodb = "2.1.0"
\ No newline at end of file
+mongodb = "2.1.0"
+percent-encoding = "2.0.0"
+thiserror = "1"
+once_cell = "1"
```
Comment (tomhoule) [inline libs/mongodb-client/Cargo.toml:12]: Nitpick, but maybe we could separate these deps to make it clear they can be removed once the io-free url parsing is available upstream

File: libs/mongodb-client/src/lib.rs [hunk 2]
```diff
@@ -1,17 +1,291 @@
...
         ClientOptions::parse(connection_string).await?
     };
 
-    Client::with_options(options)
+    Ok(Client::with_options(options)?)
+}
+
+/// The parts we need taken from `mongodb` private functions. Please remove me
+/// when they make these apis public.
+pub struct MongoConnectionString {
```
Comment (tomhoule) [inline libs/mongodb-client/src/lib.rs:26]: Nitpick, but it's not completely clear whether we can remove everything, or everything but the tests, etc. Maybe a separate file? (sorry, it's really a nitpick, feel free to ignore)

**Review by tomhoule — APPROVED** [review]

File: libs/mongodb-client/src/lib.rs (hunk 2 above)
Comment (tomhoule) [reply libs/mongodb-client/src/lib.rs:26]: We can also remove the error module when that's upstream, right?

Comment (pimeys) [reply libs/mongodb-client/src/lib.rs:26]: Yep.

File: libs/mongodb-client/Cargo.toml (hunk 1 above)
Comment (pimeys) [reply libs/mongodb-client/Cargo.toml]: Yep.

**Pull request merged**